File: backend/src/main/java/com/cleanviewai/backend/service/ChatService.java
package com.cleanviewai.backend.service;

import com.cleanviewai.backend.dto.request.ChatMessageRequest;
import com.cleanviewai.backend.dto.response.ChatMessageResponse;
import com.cleanviewai.backend.entity.ChatMessage;
import com.cleanviewai.backend.entity.ChatSession;
import com.cleanviewai.backend.repository.ChatMessageRepository;
import com.cleanviewai.backend.repository.ChatSessionBlockRepository;
import com.cleanviewai.backend.repository.ChatSessionRepository;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestClient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class ChatService {

    private final RestClient aiEngineClient;
    private final ChatSessionRepository chatSessionRepository;
    private final ChatMessageRepository chatMessageRepository;
    private final ChatSessionBlockRepository chatSessionBlockRepository;

    @Transactional
    public ChatMessageResponse sendMessage(String email, ChatMessageRequest req) {
        ChatSession session = resolveSession(email, req);

        // 2차 맥락 필터: 유저의 모든 세션 ID
        List<Long> allowedSessionIds = chatSessionRepository
                .findByUserIdOrderByUpdatedAtDesc(email)
                .stream()
                .map(ChatSession::getId)
                .toList();

        // 차단 필터: 현재 세션에서 차단된 과거 세션 ID
        List<Long> excludedSessionIds = chatSessionBlockRepository
                .findBlockedSessionIdsByCurrentSessionId(session.getId());

        saveMessage(session, "user", req.message());

        // FastAPI 요청 바디
        Map<String, Object> body = new HashMap<>();
        body.put("message", req.message());
        body.put("user_id", email);
        body.put("session_id", session.getId());
        body.put("allowed_session_ids", allowedSessionIds);
        body.put("excluded_session_ids", excludedSessionIds);

        // OTel W3C Trace Context 헤더 주입 (분산 추적 계보 연결)
        String traceparent = buildTraceparent();

        AiEngineResponse aiResponse = aiEngineClient.post()
                .uri("/api/v1/chat/ask")
                .header("traceparent", traceparent)
                .body(body)
                .retrieve()
                .body(AiEngineResponse.class);

        saveMessage(session, "assistant", aiResponse.answer());
        session.touch();

        // AI 응답을 가공 없이 그대로 프론트로 전달 (session_id만 추가)
        return new ChatMessageResponse(
                aiResponse.answer(),
                session.getId().toString(),
                aiResponse.traceId(),
                aiResponse.segments(),
                aiResponse.references()
        );
    }

    // ── private helpers ──────────────────────────────────────────────────────

    private ChatSession resolveSession(String email, ChatMessageRequest req) {
        if (req.sessionId() != null) {
            try {
                Long sessionId = Long.parseLong(req.sessionId());
                return chatSessionRepository.findByIdAndUserId(sessionId, email)
                        .orElseGet(() -> createSession(email, req.message()));
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("유효하지 않은 세션 ID입니다: " + req.sessionId());
            }
        }
        return createSession(email, req.message());
    }

    private ChatSession createSession(String email, String firstMessage) {
        String title = firstMessage.length() > 20
                ? firstMessage.substring(0, 20)
                : firstMessage;
        return chatSessionRepository.save(ChatSession.builder()
                .userId(email)
                .title(title)
                .build());
    }

    private void saveMessage(ChatSession session, String role, String content) {
        chatMessageRepository.save(ChatMessage.builder()
                .session(session)
                .role(role)
                .content(content)
                .build());
    }

    /**
     * W3C Trace Context 표준 traceparent 헤더 값 생성
     * 포맷: 00-{traceId}-{spanId}-01
     */
    private String buildTraceparent() {
        SpanContext ctx = Span.current().getSpanContext();
        if (ctx.isValid()) {
            return String.format("00-%s-%s-01", ctx.getTraceId(), ctx.getSpanId());
        }
        // OTel이 비활성화된 로컬 환경에서는 NOOP 스판이 반환되므로 기본값 사용
        return "00-00000000000000000000000000000000-0000000000000000-00";
    }

    // ── AI 엔진 응답 DTO ─────────────────────────────────────────────────────

    private record AiEngineResponse(
            String answer,
            @JsonProperty("trace_id") String traceId,
            List<ChatMessageResponse.SegmentDto> segments,
            List<ChatMessageResponse.ReferenceDto> references
    ) {}
}

File: backend-rust/src/service/chat_service.rs
use serde::{Deserialize, Serialize};

use crate::dto::request::chat::ChatMessageRequest;
use crate::dto::response::chat::{ChatMessageResponse, ReferenceDto, SegmentDto};
use crate::error::AppError;
use crate::repository::chat_message_repo::ChatMessageRepository;
use crate::repository::chat_session_block_repo::ChatSessionBlockRepository;
use crate::repository::chat_session_repo::ChatSessionRepository;

pub struct ChatService;

#[derive(Debug, Serialize)]
struct AiEngineRequest {
    message: String,
    user_id: String,
    session_id: i64,
    allowed_session_ids: Vec<i64>,
    excluded_session_ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
struct AiEngineResponse {
    answer: String,
    trace_id: Option<String>,
    segments: Option<Vec<SegmentDto>>,
    references: Option<Vec<ReferenceDto>>,
}

impl ChatService {
    pub async fn send_message(
        pool: &sqlx::PgPool,
        email: &str,
        req: ChatMessageRequest,
        ai_engine_url: &str,
    ) -> Result<ChatMessageResponse, AppError> {
        let session = Self::resolve_session(pool, email, &req).await?;

        let allowed_sessions = ChatSessionRepository::find_by_user_id(pool, email).await?;
        let allowed_session_ids: Vec<i64> = allowed_sessions.iter().map(|s| s.id).collect();

        let excluded_session_ids = ChatSessionBlockRepository::find_blocked_session_ids(pool, session.id).await?;

        ChatMessageRepository::create(pool, session.id, "user", &req.message).await?;

        let ai_req = AiEngineRequest {
            message: req.message,
            user_id: email.to_string(),
            session_id: session.id,
            allowed_session_ids,
            excluded_session_ids,
        };

        let client = reqwest::Client::new();
        let ai_resp: AiEngineResponse = client
            .post(format!("{}/api/v1/chat/ask", ai_engine_url))
            .json(&ai_req)
            .send()
            .await?
            .json()
            .await?;

        ChatMessageRepository::create(pool, session.id, "assistant", &ai_resp.answer).await?;
        ChatSessionRepository::touch(pool, session.id).await?;

        Ok(ChatMessageResponse {
            answer: ai_resp.answer,
            session_id: session.id.to_string(),
            trace_id: ai_resp.trace_id,
            segments: ai_resp.segments,
            references: ai_resp.references,
        })
    }

    async fn resolve_session(
        pool: &sqlx::PgPool,
        email: &str,
        req: &ChatMessageRequest,
    ) -> Result<crate::model::ChatSession, AppError> {
        if let Some(ref session_id_str) = req.session_id {
            let parsed: Result<i64, _> = session_id_str.parse();
            if let Ok(session_id) = parsed {
                if let Some(session) = ChatSessionRepository::find_by_id_and_user_id(pool, session_id, email).await? {
                    return Ok(session);
                }
            }
        }
        Self::create_session(pool, email, &req.message).await
    }

    async fn create_session(
        pool: &sqlx::PgPool,
        email: &str,
        first_message: &str,
    ) -> Result<crate::model::ChatSession, AppError> {
        let title = if first_message.len() > 20 {
            &first_message[..20]
        } else {
            first_message
        };

        Ok(ChatSessionRepository::create(pool, email, title).await?)
    }
}
